Docker deployment and documentation: various improvements

Theodlz opened the pull request:
In this PR, we:
- move from an ubuntu base image to a lighter debian "slim" base image
- reorganize and group steps of the Dockerfile to minimize the number of layers
- change the port mapping from the docker-compose.yaml file which mapped the internal port 9000 to 5000, when the app is already running on 5000 in said container.
- add a mention of the port in the `docker.yaml` file, to make it clear how to change the ports on which the app is running and the external facing port (useful for webhooks for example).
- update the deployment documentation to make things clearer for others that want to deploy skyportal with docker/k8
- fix a minor bug I noticed while working on this PR, with the observation plan auto-triggering feature.

stefanv commented on doc/deploy.md:
```diff
@@ -12,58 +12,86 @@ same information applies when using Kubernetes, for which we have an
-docker exec skyportal-web-1 bash -c 'source /skyportal_env/bin/activate && FLAGS="--create_tables --config=config.yaml" make load_demo_data'
+This will mount the `config.yaml` file from the `config` directory to the `/etc/skyportal/config.yaml` path in the container. You could technically directly have it mounted to `/skyportal/config.yaml`, but we recommend using a different path to not lose track of the original configuration file used to build the image.
+
+You would need to pass the `--config=config.yaml` flag when calling any `make` commands. For the container to use this configuration file on startup, you can simply specify your own FLAG environment variable in the `docker-compose.yaml` file:
```
```suggestion
You would need to pass the `--config=config.yaml` flag when calling any `make` commands. For the container to use this configuration file on startup, set the FLAG environment variable in the `docker-compose.yaml` file:
```

stefanv commented on docker-compose.yaml:
```diff
@@ -4,8 +4,8 @@ services:
     ports:
-      - "9000:5000"
-      - "9001:5001"
+      - "5000:5000"
```
Isn't this supposed to map internal 5000 to external 9000, so docker images can run on macos?

Theodlz replied:
Maybe, in which case the existing config is incorrect then? The docker ports should be <internal>:<external>, right?

Theodlz replied:
Or maybe I got it reversed, might be ...

stefanv replied:
Can never remember :)

Theodlz replied:
yep its external -> internal, so I reversed that change and tested it

Theodlz commented on skyportal/handlers/api/observation_plan.py:
```diff
@@ -263,7 +263,10 @@ def send_observation_plan(plan_id, session, auto_send=False, default_obsplan_id=
         # if the plan's end date is in the past (can see that in the payload), we skip the auto-send
         plan_request_end_date = observation_plan_request.payload.get("end_date", None)
         if plan_request_end_date:
-            if arrow.get(plan_request_end_date).datetime < datetime.utcnow():
```
Sure thing

Theodlz replied:
Removed

Theodlz commented on Dockerfile:
```diff
@@ -1,53 +1,42 @@
-
-ENV LC_ALL=C.UTF-8
-ENV LANG=C.UTF-8
+# RUN git clone https://github.com/Theodlz/snid-install-ubuntu.git && \
```
That's on purpose. We want this to be an option that some advanced user can consider. I mean I'm fine removing it since we don't use it in the app.

stefanv replied:
Ah, OK, would need a comment to explain what it is then!

Theodlz replied:
Ended up removing it!

Theodlz commented on Dockerfile:
```diff
@@ -1,53 +1,42 @@
+    python3 -m venv /skyportal_env && \
     source /skyportal_env/bin/activate && \
-    export NPM_CONFIG_LEGACY_PEER_DEPS=true && \
+    pip install --upgrade pip wheel packaging setuptools --no-cache && \
```
so we don't start filling the pip cache at all. We anyway flush it later on but we might as well not populate it at all.

Theodlz commented on Makefile:
```diff
@@ -50,12 +55,12 @@ prepare_seed_data:
 load_demo_data: ## Import example dataset
 load_demo_data: FLAGS := $(if $(FLAGS),$(FLAGS),--config=config.yaml)
-load_demo_data: | dependencies prepare_seed_data
+load_demo_data: | dependencies_no_js prepare_seed_data
```
we don't need them to load demo data in the backend. In the Dockerfile we drop the `node_modules` after building the bundle, but then if we run `make load_demo_data` to create the tables and such (as instructed), it takes forever and uses a lot of disk space because it recreates all of the js dependencies

stefanv replied:
Oooh I see! Yes, makes sense.

Theodlz commented on doc/deploy.md:
````diff
@@ -12,58 +12,86 @@ same information applies when using Kubernetes, for which we have an
-```
-
-Otherwise, you may build the images from scratch:
+After customizing your `docker.yaml` file (the equivalent of `config.yaml`, used as the config file when building a docker image), you may build the images using the following command:
````
I don't think so! It's literally what it is. We copy it as config.yaml in the Dockerfile! I think that on the other hand if we don't explain that, folks won't necessarily understand that they need to edit this file when building the image and not config.yaml

stefanv replied:
Sorry, thought you were talking of Docker file, now see you're talking about config used in Docker. I wonder how we can make this even clearer.

Theodlz commented on doc/deploy.md:
```diff
@@ -12,58 +12,86 @@ same information applies when using Kubernetes, for which we have an
+Note that by default, the SkyPortal image will run the application in production (`make run_production`, see the `CMD` directive at the end of the Dockerfile).
 
-Inspect the logs for the running containers using:
+The key behavior to note when running production mode is that the application will not create any database tables automatically (to avoid messing with production data). This means that the very first time you spin up SkyPortal using containers, the `skyportal_dbdata` volume will only have an empty database. While the `skyportal-web-1` container will start, the web application will not be accessib ...
```
Not sure what is the command for just initialize and empty DB, but I'll check

stefanv replied:
Won't a lot of people need that, if they want to start fresh?

Theodlz replied:
They would! I added a make command to create the tables.

Theodlz commented:
@stefanv I addressed all of your comments. Please let me know if you can review the current state of the PR, I'd love to merge this soon-ish, as some folks would greatly benefit from the detailed instructions

stefanv commented on Dockerfile:
```diff
@@ -1,53 +1,38 @@
+ENV NODE_MAJOR=20
+ENV NPM_CONFIG_LEGACY_PEER_DEPS=true
+ENV PATH="/root/.cargo/bin:${PATH}"
+ENV SNCOSMO_DATA_DIR=/skyportal/persistentdata/sncosmo
```
Leftover?

stefanv commented on Dockerfile:
```diff
@@ -1,53 +1,38 @@
+ENV LANG=C.UTF-8
+ENV NODE_MAJOR=20
+ENV NPM_CONFIG_LEGACY_PEER_DEPS=true
+ENV PATH="/root/.cargo/bin:${PATH}"
```
Necessary?

Theodlz replied:
I can drop the node & npm stuff, bt will keep adding cargo to the bath and the SNCOSMO env var, it can get confused without that

Theodlz replied:
I take that back we need all of that, I thought I was on the bun PR for some reason, we do need all these vars

stefanv commented on Dockerfile:
```diff
@@ -69,12 +54,17 @@ RUN bash -c "\
 
 USER skyportal
 
+# edit the exposed port to match the one in the config.yaml
```
docker.yaml?

Theodlz replied:
Ah right, thanks!

stefanv reviewed (approved):
Left some minor comments, but feel free to merge once you're ready.
